src/GafferScene/InteractiveRender.cpp

```diff
@@ -217,6 +286,437 @@ void InteractiveRender::parentChanged( GraphComponent *child, GraphComponent *ol
+		{
+			ContextPtr context = new Context( *m_context, Context::Borrowed );
+			context->set( ScenePlug::scenePathContextName, m_scenePath );
+			Context::Scope scopedContext( context.get() );
+
+			m_sceneGraph->m_attributes = m_scene->attributesPlug()->getValue();
+			const BoolData *visibilityData = m_sceneGraph->m_attributes->member<BoolData>( SceneInterface::visibilityName );
+			if( visibilityData && !visibilityData->readable() )
+			{
+				// terminate recursion for invisible locations
+				return NULL;
+			}
+			
+			// compute child names:
+			IECore::ConstInternedStringVectorDataPtr childNamesData = m_scene->childNamesPlug()->getValue();
+
+			const std::vector<IECore::InternedString> &childNames = childNamesData->readable();
+			if( childNames.empty() )
+			{
+				// nothing more to do
+				return NULL;
+			}
+
+			// add children for this location:
+			std::vector<InteractiveRender::SceneGraph *> children;
+			for( std::vector<IECore::InternedString>::const_iterator it = childNames.begin(), eIt = childNames.end(); it != eIt; ++it )
+			{
+				SceneGraph *child = new SceneGraph();
+				child->m_name = *it;
+				child->m_parent = m_sceneGraph;
+				children.push_back( child );
+			}
+
+			// spawn child tasks:
+			set_ref_count( 1 + children.size() );
+			ScenePlug::ScenePath childPath = m_scenePath;
+			childPath.push_back( IECore::InternedString() ); // space for the child name
+			for( std::vector<SceneGraph *>::const_iterator it = children.begin(), eIt = children.end(); it != eIt; ++it )
+			{
+				childPath.back() = (*it)->m_name;
+				SceneGraphBuildTask *t = new( allocate_child() ) SceneGraphBuildTask(
+					m_scene,
+					m_context,
+					(*it),
+					childPath
+				);
+				
+				spawn( *t );
+			}
+
+			wait_for_all();
+			
+			// add visible children to m_sceneGraph->m_children:
+			for( std::vector<SceneGraph *>::const_iterator it = children.begin(), eIt = children.end(); it != eIt; ++it )
+			{
+				const BoolData *visibilityData = (*it)->m_attributes->member<BoolData>( SceneInterface::visibilityName );
+				if( visibilityData && !visibilityData->readable() )
+				{
+					continue;
+				}
+				m_sceneGraph->m_children.push_back( *it );
+			}
+			
+			return NULL;
+		}
+
+	private :
+
+		const ScenePlug *m_scene;
+		const Context *m_context;
+		SceneGraph *m_sceneGraph;
+		ScenePlug::ScenePath m_scenePath;
+};
+
+
+//////////////////////////////////////////////////////////////////////////
+// SceneGraphIteratorFilter implementation
+//
+// Does a serial, depth first traversal of a SceneGraph hierarchy based at
+// "start", and spits out SceneGraph* tokens:
+//
+//////////////////////////////////////////////////////////////////////////
+
+class InteractiveRender::SceneGraphIteratorFilter : public tbb::filter
+{
+	public:
+		SceneGraphIteratorFilter( InteractiveRender::SceneGraph *start ) :
+			tbb::filter( tbb::filter::serial_in_order ), m_current( start )
+		{
+			m_childIndices.push_back( 0 );
+		}
+		
+		void next()
```

Comment:
I think I'm right in saying that this is an implementation detail of SceneGraphIteratorFilter? If that's the case, can you make it private to make it clear that it's not something that tbb will use directly.